README.md
# okayNowThis.js
A jQuery plugin that makes pausing and resuming queued functions a cinch.

[View the demo](https://jessengatai.github.io/okayNowThis.js/)

## Getting Started
The only requirement for okayNowThis.js to function is jQuery, so be sure to call that before the plugin.

```html
<head>
	<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
	<script type="text/javascript" src="okayNowThis.min.js"></script>
</head>
```

## Basic Usage
The plugin works by manipulating queues attached to DOM elements that are built using jQuery functions like 'delay'. Meaning you will need to call okayNowThis() on an element that has an unique ID. Throughout the docs we will refer to this element as the 'staging' element.

```javascript
$('#yourstage').okayNowThis(method,delay,function(){
	// some code to run
});
```

## Methods
There are 4 methods you can pass to the plugin.
* keyframe
* pause
* resume
* reset

#### Keyframe
Each keyframe will be appended onto the queue attached to your staging element.

The keyframe method requires us to pass a delay in milliseconds. This delay dictates the time in takes before the next keyframe fires. This is ideal if your adding or removing a CSS class with a transition time, as we can make sure the CSS animation finishes before moving onto the next keyframe.

```javascript
$('.play').on('click',function(){

	// keyframe 1
	$('#yourstage').okayNowThis('keyframe',0,function(){
		// code to run with no delay, like show our pause button!

	// keyframe 2
	}).okayNowThis('keyframe',1000,function(){
		// code to run with 1 second delay after keyframe 1

	// keyframe 3
	}).okayNowThis('keyframe',500,function(){
		// code to run with half second delay after keyframe 2

	});

});
```

#### Pause
Does what it says on the box. Works by cloning the queue attached to the staging element so we can re-attached later.
```javascript
$('.pause').on('click',function(){
	$('#yourstage').okayNowThis('pause',function(){
		// do something on pause, like show our resume button!
	});
});
```

#### Resume
Resume will re-attach the queue and re-add the plugin classes to the staging element.
```javascript
$('.resume').on('click',function(){
	$('#yourstage').okayNowThis('resume',function(){
		// do something on resume, like show our pause button!
	});
});
```

#### Reset
Reset will delete the queue and remove the plugin classes attached to the staging element.
```javascript
$('.reset').on('click',function(){
	$('#yourstage').okayNowThis('reset',function(){
		// do something on reset, like clean up all the mess we probably made!
	});
});
```

## Classes
There are 3 classes added to the staging element
* okayNowThis-playing
* okayNowThis-paused
* keyframe-x (where x is the current keyframe number)


CHANGELOG.md
# 1.0.1

2018-03-12

 - camelCase the plugin name
 - update the repo structure & organise files
 - give the demo page styles a bit of an overhaul

# 1.0.0

2016-10-12

 - initial release


scripts/main.js
jQuery(document).ready(function($) {

	// logs
	const body = $('body');
	const log_c = $('#log');
	const log_z = $('#status');
	const log_x = $('#keyframe');

	// buttons
	const btn_pause = $('button.pause');
	const btn_resume = $('button.resume');
	const btn_reset = $('button.reset');
	const btn_play = $('button.play');

	// geometry
	const shape1 = $('#shape-1');
	const shape2 = $('#shape-2');
	const shape3 = $('#shape-3');

	/**
	 * Setup pause button
	 */
	btn_pause.on('click',function(){
		$('#stage').okayNowThis('pause',function(){
			btn_resume.removeAttr('disabled');
			btn_pause.attr('disabled','disabled');
			log_c.prepend('paused <br>');
			log_z.text('.okayNowThis-paused');
		});
	});

	/**
	 * Setup resume button
	 */
	btn_resume.on('click',function(){
		$('#stage').okayNowThis('resume',function(){
			btn_pause.removeAttr('disabled');
			btn_resume.attr('disabled','disabled');
			log_c.prepend('resumed <br>');
			log_z.text('.okayNowThis-playing');
		});
    console.log('this is a log');
	});

	/**
	 * Setup reset button
	 */
	btn_reset.on('click',function(){
		reset_animation();
	});

	/**
	 * Setup play button
	 */
	btn_play.on('click',function(){

		// disable play button
		btn_play.attr('disabled','disabled');
		// enable the pause button
		btn_pause.removeAttr('disabled');
		// add to logs
		log_z.text('.okayNowThis-playing');

		/**
		 * We are declaring stage inside local scope because if the reset
		 * button is pressed the old stage variables is removed from the dom
		 */
		let stage = $('#stage');

		// build our queue
		stage.okayNowThis('keyframe',0,function(){
			stage.css('width','10%');
			shape1.addClass('angle');
			log_c.prepend('function 1: 0ms delay </br>');
			log_x.text('.keyframe-1');

		}).okayNowThis('keyframe',1000,function(){
			stage.css('width','20%');
			log_c.prepend('function 2: 1000ms delay </br>');
			log_x.text('.keyframe-2');

		}).okayNowThis('keyframe',500,function(){
			stage.css('width','30%');
			body.addClass('rain-on');
			log_c.prepend('function 3: 500ms delay </br>');
			log_x.text('.keyframe-3');

		}).okayNowThis('keyframe',1000,function(){
			stage.css('width','40%');
			body.addClass('triangle-in');
			log_c.prepend('function 4: 1000ms delay </br>');
			log_x.text('.keyframe-4');

		}).okayNowThis('keyframe',1500,function(){
			stage.css('width','50%');
			log_c.prepend('function 5: 1500ms delay </br>');
			log_x.text('.keyframe-5');

		}).okayNowThis('keyframe',700,function(){
			shape1.addClass('circle');
			stage.css('width','60%');
			log_c.prepend('function 6: 700ms delay </br>');
			log_x.text('.keyframe-6');

		}).okayNowThis('keyframe',1000,function(){
			stage.css('width','70%');
			body.addClass('block-in');
			body.addClass('triangle-out');
			log_c.prepend('function 7: 1000ms delay </br>');
			log_x.text('.keyframe-7');

		}).okayNowThis('keyframe',600,function(){
			stage.css('width','80%');
			body.addClass('vertical-rain-on');
			log_c.prepend('function 8: 600ms delay </br>');
			log_x.text('.keyframe-8');

		}).okayNowThis('keyframe',1000,function(){
			stage.css('width','90%');
			shape1.removeClass('angle circle');
			log_c.prepend('function 9: 1000ms delay </br>');
			log_x.text('.keyframe-9');

		}).okayNowThis('keyframe',700,function(){
			stage.css('width','100%');
			body.removeClass('angle circle');
			body.addClass('block-out');
			log_c.prepend('function 10: 700ms delay </br>');
			log_x.text('.keyframe-10');

		}).okayNowThis('keyframe',800,function(){
			stage.css('width','1%');
			log_c.prepend('function 11: 800ms delay </br>');
			log_x.text('.keyframe-11');

		}).okayNowThis('keyframe',600,function(){
			log_c.prepend('function 12: 600ms delay </br>');
			log_x.text('.keyframe-12');

			reset_animation(); // just for good measure we are reseting the entire stage at the end

		});
		// end of queue

	});

	// our reset functions
	function reset_animation(){

		// we are declaring stage again because if the reset button is pressed the old stage variables are useless
		var stage = $('#stage');
		var wrap = $('#stage-wrap');
		var shape1 = $('#shape-1');

		stage.okayNowThis('reset',function(){
			// delete the entire staging element
			stage.remove();
			// add the original stage markup back in (for complex markup consider using .clone on document ready)
			wrap.append('<div id="stage" class="progress-bar"></div>');
			// clean up any body classes that may have been added during the slideshow
			body.removeClass();
			shape1.removeClass();
			// clean up our buttons
			btn_pause.attr('disabled','disabled');
			btn_resume.attr('disabled','disabled');
			btn_play.removeAttr('disabled');
			// add our message
			log_c.prepend('reset (removes plugin classes and deletes the queue) <br>');
			log_z.text('');
			log_x.text('');
		});

	}

});
